datafusion/src/execution/memory_manager.rs

```diff
@@ -0,0 +1,320 @@
+   │                                              z                             │
+   │                                              z                             │
+   └──────────────────────────────────────────────z─────────────────────────────┘
+*/
+
+/// Manage memory usage during physical plan execution
+pub struct MemoryManager {
+    consumers: Arc<Mutex<HashMap<MemoryConsumerId, Arc<dyn MemoryConsumer>>>>,
+    trackers: Arc<Mutex<HashMap<MemoryConsumerId, Arc<dyn MemoryConsumer>>>>,
+    trackers_total_usage: AtomicUsize,
+    pool_size: usize,
+    join_handle: Arc<Mutex<Option<JoinHandle<()>>>>,
+}
+
+impl MemoryManager {
+    /// Create new memory manager based on max available pool_size
+    pub fn new(pool_size: usize) -> Self {
+        info!(
+            "Creating memory manager with initial size {}",
+            human_readable_size(pool_size)
+        );
+        Self {
+            consumers: Arc::new(Mutex::new(HashMap::new())),
+            trackers: Arc::new(Mutex::new(HashMap::new())),
+            trackers_total_usage: AtomicUsize::new(0),
+            pool_size,
+            join_handle: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    fn update_tracker_total(self: &Arc<Self>) {
+        let trackers = self.trackers.lock().unwrap();
+        if trackers.len() > 0 {
+            let sum = trackers.values().fold(0usize, |acc, y| acc + y.mem_used());
+            drop(trackers);
+            self.trackers_total_usage.store(sum, Ordering::SeqCst);
+        }
+    }
+
+    /// Initialize
+    pub(crate) fn initialize(self: &Arc<Self>) {
+        let manager = self.clone();
+        let handle = task::spawn(async move {
+            let mut interval_timer = tokio::time::interval(Duration::from_secs(60));
+            loop {
+                interval_timer.tick().await;
+                manager.update_tracker_total();
+            }
+        });
+        let _ = self.join_handle.lock().unwrap().insert(handle);
+    }
+
+    /// Register a new memory consumer for memory usage tracking
+    pub(crate) fn register_consumer(self: &Arc<Self>, consumer: Arc<dyn MemoryConsumer>) {
+        let id = consumer.id().clone();
+        match consumer.type_() {
+            ConsumerType::Controlling => {
+                let mut consumers = self.consumers.lock().unwrap();
+                consumers.insert(id, consumer);
+            }
+            ConsumerType::Tracking => {
+                let mut trackers = self.trackers.lock().unwrap();
+                trackers.insert(id, consumer);
+            }
+        }
+    }
+
+    /// Grow memory attempt from a consumer, return if we could grant that much to it
+    async fn try_grow(
+        self: &Arc<Self>,
+        required: usize,
+        current: usize,
+        consumer_id: &MemoryConsumerId,
+    ) -> bool {
+        let max_per_op = {
+            let total_available =
+                self.pool_size - self.trackers_total_usage.load(Ordering::SeqCst);
```

Comment:
due to background refresh delay, this is best memory usage counting right? if so, it's worth documenting.